shotover-proxy/src/transforms/redis/sink_cluster.rs

```diff
@@ -547,23 +565,29 @@ enum ChannelResult {
 
 impl RoutingInfo {
     #[inline(always)]
-    pub fn for_command_frame(args: &[RedisFrame]) -> Result<Option<RoutingInfo>> {
+    pub fn for_command_frame(args: &[RedisFrame]) -> Result<RoutingInfo> {
         let command_name = match args.get(0) {
             Some(RedisFrame::BulkString(command_name)) => command_name.to_ascii_uppercase(),
             _ => bail!("syntax error: bad command name"),
         };
 
         Ok(match command_name.as_slice() {
-            b"FLUSHALL" | b"FLUSHDB" => Some(RoutingInfo::AllMasters),
+            b"FLUSHALL" | b"FLUSHDB" => RoutingInfo::AllMasters(ResponseJoin::First),
             b"ECHO" | b"CONFIG" | b"CLIENT" | b"SLOWLOG" | b"DBSIZE" | b"LASTSAVE" | b"PING"
-            | b"INFO" | b"BGREWRITEAOF" | b"BGSAVE" | b"SAVE" | b"TIME" | b"KEYS" | b"ACL" => {
-                Some(RoutingInfo::AllNodes)
+            | b"INFO" | b"BGREWRITEAOF" | b"BGSAVE" | b"SAVE" | b"TIME" | b"ACL" => {
+                RoutingInfo::AllNodes
             }
+
+            b"KEYS" => RoutingInfo::AllMasters(ResponseJoin::ArrayJoin),
             b"SCRIPT" => match args.get(1) {
-                Some(RedisFrame::BulkString(a)) if a.to_ascii_uppercase() == b"KILL" => None,
-                _ => Some(RoutingInfo::AllMasters),
+                Some(RedisFrame::BulkString(a)) if a.to_ascii_uppercase() == b"KILL" => {
+                    RoutingInfo::Unsupported
+                }
+                _ => RoutingInfo::AllMasters(ResponseJoin::First),
             },
-            b"SCAN" | b"SHUTDOWN" | b"SLAVEOF" | b"REPLICAOF" | b"MOVE" | b"BITOP" => None,
+            b"SCAN" | b"SHUTDOWN" | b"SLAVEOF" | b"REPLICAOF" | b"MOVE" | b"BITOP" => {
+                RoutingInfo::Unsupported
+            }
             b"EVALSHA" | b"EVAL" => match args.get(2) {
                 Some(RedisFrame::BulkString(key_count)) => std::str::from_utf8(key_count)
                     .unwrap_or("0")
```

Comment:
Not your changes, the `key_count` variable below this should be a `u64`

Reply:
Hmmmm I cant imagine ever needing that bit https://redis.io/commands/eval/

But maybe we should do a string check against "0" instead anyway?

Reply:
I'll just make it a u64 and leave it at that :shrug: the current logic looks wrong (why do we only use one key?) and will probably need to be redone later on.

Reply:
from a routing perspective eval commands get routed on the first key passed into the lua script

let me find the doc ref that specifies this

Reply:
If thats true then we can write a much simpler implementation:
```rust
            b"EVALSHA" | b"EVAL" => match args.get(2) {
                Some(RedisFrame::BulkString(key_count)) => {
                    if key_count.as_ref() == b"0" {
                        RoutingInfo::Random
                    } else {
                        args.get(3)
                            .and_then(RoutingInfo::for_key)
                            .unwrap_or(RoutingInfo::Unsupported)
                    }
                }
                _ => RoutingInfo::Unsupported,
            },
```

Reply:
oh cool, redis doesnt accept 000 so we dont care
![image](https://user-images.githubusercontent.com/5120858/165653842-7955078d-aba0-4710-8278-edc670c771da.png)

Reply:
> why do we only use one key?

Lua scripts on Redis is normally only allowed to access data that is in the same ~node~ slot, so I guess it was a shortcut to only consider the first key and defer the validation to Redis. Although after digging deeper, this is more complicated... If you declare all the keys upfront as you're supposed to, Redis will return CROSSSLOT errors if the keys don't all hash to the same slot. But if you access undeclared keys that are cross slots, Redis v6 will allow that. This behaviour has only just been disabled by default on Redis v7 (redis/redis#10615).

This doesn't change the fact that "Redis scripts block access to non-local keys", so for Shotover, it is probably a reasonable shortcut to keep? Is there any benefit with pre-validating if Redis will do it again? :/

https://github.com/redis/redis/blob/efcd1bf394668e418df1a93cd28cf9e8b0c09ce5/src/script.c#L399
https://github.com/redis/redis/blob/3cd8baf61610416aab45e0bcedcaab9beae80184/src/cluster.c#L6677

> should do a string check against "0" instead anyway?

Is this referring to the `unwrap_or("0")`? 🤔 The key_count should only be an ASCII number. If we can't parse it as UTF-8, it's a protocol error. I think we should just abort and return an error back to the client?

Reply:
ooooh thanks @XA21X that explanation makes a lot of sense, I will go ahead and use the code snippet I posted earlier.

> Is this referring to the unwrap_or("0")? thinking The key_count should only be an ASCII number. If we can't parse it as UTF-8, it's a protocol error. I think we should just abort and return an error back to the client?

I just meant that I will do the check directly against the string bytes rather than parsing to an int.